### PR TITLE
Allow providers which are selected to be installed in place of existing packages which do not satisfy requirements

### LINKED ISSUE #6753: Provide does not satisfy package requirement if package with same name exists
I believe this should be allowed. It works if you swap replace/provide.

```
{
	"repositories": [
		{
			"packagist": false
		},
		{
			"type": "package",
			"package": [
				{
					"name": "foo/polyfill",
					"provide": {
						"foo/standard": "1.0.0"
					},
					"replace": {
					},
					"version": "1.0.0"
				},
				{
					"name": "foo/standard",
					"require": {
						"foo/does-not-exist": "1.0.0"
					},
					"version": "1.0.0"
				}
			],
			"type": "package"
		}
	],
	"require": {
		"foo/standard": "1.0.0",
		"foo/polyfill": "1.0.0"
	}
}
```

### PR DESCRIPTION
**Pull request by Seldaek**
Fixes #6753

I think it's fine really as per the 3 new tests, as they only get installed if selected, this isn't risky.

### CODE REVIEW

File: src/Composer/DependencyResolver/Pool.php [hunk 1]
```diff
@@ -147,11 +147,6 @@ private function computeWhatProvides($name, $constraint, $mustMatchName = false)
             }
         }
 
-        // if a package with the required name exists, we ignore providers
-        if ($nameMatch) {
```
Comment (stof) [inline src/Composer/DependencyResolver/Pool.php]: Computing `$nameMatch` is now useless, as it is never read. so it should be removed.

File: tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-if-selected.test [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "repositories": [
+        {
+            "type": "package",
+            "package": [
+                {
+                    "name": "foo/polyfill",
+                    "provide": {
+                        "foo/standard": "1.0.0"
+                    },
+                    "replace": {
```
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-if-selected.test]: can be dropped?

File: tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-if-selected.test [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                },
+                {
+                    "name": "foo/standard",
+                    "require": {
+                        "foo/does-not-exist": "1.0.0"
+                    },
+                    "version": "1.0.0"
+                }
+            ],
+            "type": "package"
```
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-if-selected.test]: duplicate of type: package at the top?

File: tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-together-with-provided-if-both-installable.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                    "replace": {
+                    },
+                    "version": "1.0.0"
+                },
+                {
+                    "name": "foo/standard",
+                    "version": "1.0.0"
+                }
+            ],
+            "type": "package"
```
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/provider-packages-can-be-installed-together-with-provided-if-both-installable.json]: same as above test: duplicate type package and unnecessary replace

**Pull request merged**